codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -16,329 +16,515 @@
+        connection().addListener(new ConnectionListener());
+        streamKey = connection().newKey();
+        initialLocalConnectionWindow = initialSettings.initialWindowSize();
+        gracefulShutdownTimeoutMillis(gracefulShutdownTimeoutMillis);
     }
 
-    Http2ConnectionHandler connectionHandler() {
-        return http2Handler;
+    private Http2FrameCodec(boolean server, Http2FrameWriter frameWriter, Http2FrameLogger frameLogger,
+                    Http2Settings initialSettings) {
+        super(server, frameWriter, frameLogger, initialSettings);
+
+        decoder().frameListener(new FrameListener());
+        connection().addListener(new ConnectionListener());
+        streamKey = connection().newKey();
+        initialLocalConnectionWindow = initialSettings.initialWindowSize();
+    }
+
+    /**
+     * Creates a new outbound/local stream.
+     */
+    DefaultHttp2FrameStream newStream() {
+        assert ctx.executor().inEventLoop();
+
+        return new DefaultHttp2FrameStream(connection());
+    }
+
+    /**
+     * Iterates over all active HTTP/2 streams.
+     *
+     * <p>This method must not be called outside of the event loop.
+     */
+    final void forEachActiveStream(final Http2FrameStreamVisitor streamVisitor) throws Http2Exception {
+        assert ctx.executor().inEventLoop();
+
+        connection().forEachActiveStream(new Http2StreamVisitor() {
+            @Override
+            public boolean visit(Http2Stream stream) {
+                Http2FrameStream stream2 = stream.getProperty(streamKey);
+                if (stream2 == null) {
+                    /**
+                     * This code is expected to almost never execute. However, in rare cases it's possible that a
+                     * stream is active without a {@link Http2FrameStream} object attached, as it's set in a listener of
+                     * the HEADERS frame write.
+                     */
+                    // TODO(buchgr): Remove once Http2Stream2 and Http2Stream are merged.
+                    return true;
+                }
+                try {
+                    return streamVisitor.visit(stream2);
+                } catch (Throwable cause) {
+                    onError(ctx, cause);
+                    return false;
+                }
+            }
+        });
     }
 
     /**
      * Load any dependencies.
      */
     @Override
-    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+    public final void handlerAdded(ChannelHandlerContext ctx) throws Exception {
         this.ctx = ctx;
-        ctx.pipeline().addBefore(ctx.executor(), ctx.name(), null, http2Handler);
-        http2HandlerCtx = ctx.pipeline().context(http2Handler);
+        super.handlerAdded(ctx);
+        handlerAdded0(ctx);
+        sendInitialConnectionWindow(ctx);
     }
 
-    /**
-     * Clean up any dependencies.
-     */
+    void handlerAdded0(@SuppressWarnings("unsed") ChannelHandlerContext ctx) throws Exception { }
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1186)